src/docfx/lib/git/FileCommitProvider.cs

```diff
@@ -168,13 +135,62 @@ public List<GitCommit> GetCommitHistory(string file, string committish = null)
             return result.Select(c => c.GitCommit).ToList();
         }
 
-        public IReadOnlyList<KeyValuePair<string, Dictionary<(long commit, long blob), (long[] commitHistory, int lruOrder)>>> BuildCache()
+        public List<GitCommit> GetDeletedFileCommitHistory(string file, int top = -1, string committish = null)
```

Comment:
instead of duplicating this complex algorithm, can we have one method that get commit history with top N and an optional cache?

Reply:
i can't think a way to get ride of `async` for optional cache, and i don't want to use `.result`, so any suggestions here?

Reply:
If could use some refactor of `Cache`, basically we can have a single method that contains the algorithm:

```csharp
List<GitCommit> GetCommitHistory(string file, int top, string committish, Cache commitCache)
```

and then have the async method calling the sync method, something similar to:

```csharp
async Task<List<GitCommit>> GetCommitHistory(string file, string committish)
{
        var cache = await GetCache();
        var result = GetCommitHistory(file, committish, cache)
        UpdateCache()
        return result;
}
```

I personally do not want changes to GitCommitProvider to mix with other changes, because there aren't enough tests in PR to cover GitCommitProvider and it is performance sensitive. (The only safeguard for GitCommitProvider right now is impact test).

Reply:
we can do the cache refactor later, current goal is to:
1. serving current purpose
2. reduce the duplicate code as much as possible
3. make the code more readable, don't change the logic at all

Reply:
we don't have any other mix changes here, just re-structure the code to make it more readable.